controllers/user.js

```diff
@@ -594,6 +685,1369 @@ function logout(req, res) {
+      error: localeError,
+    });
+  }
+
+  if (JS_CONSTANTS.REMOVE_CHECK_WAITLIST_ENABLED) {
+    const hashMatch = await checkEmailHash(account);
+
+    if (!hashMatch) {
+      const localeError = LocaleUtils.fluentFormat(
+        req.supportedLocales,
+        "remove-error-no-waitlist-match"
+      );
+      return res.status(400).json({
+        error: localeError,
+      });
+    }
+  }
+
+  const memberList = {
+    members: [
+      {
+        names: [
+          {
+            first_name: firstname,
+            middle_name: middlename ? middlename : null,
+            last_name: lastname,
+          },
+        ],
+        addresses: [
+          {
+            city: city,
+            state: state,
+            country: country,
+          },
+        ],
+        emails: [
+          {
+            email: account,
+          },
+        ],
+        birth_year: parseInt(birthyear),
+      },
+    ],
+  };
+
+  const jsonMemberList = JSON.stringify(memberList);
+
+  const memberID = await handleKanaryAPISubmission(jsonMemberList);
+  if (!user.kid) {
+    await DB.setKanaryID(user, memberID);
+  }
+
+  return res.json({ nextPage: "/user/remove-signup-confirmation" });
+}
+
+async function handleKanaryAPISubmission(memberInfo) {
+  return fetch("https://thekanary.com/partner-api/v0/accounts/", {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+      Authorization: `Bearer ${AppConstants.KANARY_TOKEN}`,
+    },
+    body: memberInfo,
+  })
+    .then((res) => res.json())
+    .then((data) => {
+      return data.id;
+    })
+    .catch((error) => {
+      console.error("there was an error posting to the api", error);
+    });
+}
+
+async function checkForEmailMatch(account, user) {
+  let emailMatch = false;
+  if (user.primary_email && user.primary_verified) {
+    if (user.primary_email === account) {
+      emailMatch = true;
+    }
+  }
+
+  const monitoredEmails = await DB.getUserEmails(user.id);
+
+  monitoredEmails.forEach((email) => {
+    if (email.email && email.verified) {
+      if (email.email === account) {
+        emailMatch = true;
+      }
+    }
+  });
+
+  return emailMatch;
+}
+
+async function checkEmailDomainMatch(account) {
+  //TODO: Remove for external testing. For internal testing, we need to check that users are signing up with a Mozilla email account
+
+  let emailDomainMatch = false;
+  if (!JS_CONSTANTS.REMOVE_CHECK_EMAIL_DOMAIN_ENABLED) {
+    //if the domain check is disabled, force a match
+    emailDomainMatch = true;
+  }
+
+  const accountDomain = account.substring(account.lastIndexOf("@") + 1);
+
+  JS_CONSTANTS.REMOVE_EMAIL_DOMAIN_LIST.forEach((emailDomain) => {
+    if (emailDomain === accountDomain) {
+      emailDomainMatch = true;
+    }
+  });
+
+  return emailDomainMatch;
+}
+
+async function getHashedWaitlist() {
+  const hashedWaitlistArray = await readFile("hashed-waitlist.txt", "binary");
+  return hashedWaitlistArray.toString().split("\n");
+}
```

Comment:
it looks like hashed-waitlist.txt is opened, read, and split on every user visit/login?  Perhaps it can be read and split once, upon server startup, and hashedWaitlistArray can be saved as a constant?

Reply:
Great catch. Didn't even think about that as an option. I'll add a ticket for this one too.

Reply:
I do a similar thing here, opening a db once on server start, and refreshing every 3 days (disregard comment saying it's every 24hrs – I need to change that):
https://github.com/mozilla/blurts-server/blob/99611af49f1ee60d3a5d92b316a86f5118b63b06/server.js#L99-L102

which calls this:
https://github.com/mozilla/blurts-server/blob/99611af49f1ee60d3a5d92b316a86f5118b63b06/ip-location-service.js#L9-L21